### PR TITLE
Enhancing C# invoker to avoid unecessary host restarts.

### PR DESCRIPTION
**Pull request by fabiocav**

### CODE REVIEW

**Comment by fabiocav** [comment]
Resolves #57

File: src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs [hunk 1]
```diff
@@ -112,10 +114,10 @@ private void ReloadScript()
                 TraceWriter.Trace(traceEvent);
             }
 
-            // If the compilation succeeded, restart the host
-            // TODO: Make this smarter so we only restart the host if the 
-            // method signature (or types used in the signature) change.
-            if (!compilationResult.Any(d => d.Severity == DiagnosticSeverity.Error))
+            // If the compilation succeeded AND our function signature has changed, restart the host
+            if (!compilationResult.Any(d => d.Severity == DiagnosticSeverity.Error) &&
+                !_functionSignature.HasLocalTypeReference &&
```
Comment (davidebbo) [inline src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs]: What does `HasLocalTypeReference` mean? Maybe capture that in the comment? It seems to disable the optimization.

File: src/WebJobs.Script/Description/CSharp/CSharpFunctionSignature.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.Azure.WebJobs.Script.Description
+{
+    /// <summary>
+    /// Provides function identity validation and identification.
+    /// </summary>
+    [CLSCompliant(false)]
+    public sealed class CSharpFunctionSignature
```
Comment (davidebbo) [inline src/WebJobs.Script/Description/CSharp/CSharpFunctionSignature.cs]: `IEquatable<CSharpFunctionSignature>`?

File: src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs (hunk 1 above)
Comment (fabiocav) [inline src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs]: This flag indicates that one of the parameters uses a type that was locally defined in the function (e.g. a POCO used for a blob trigger). I'll add a comment inline as well as XML doc to the property to make that clear.

File: src/WebJobs.Script/Description/CSharp/CSharpFunctionSignature.cs (hunk 2 above)
Comment (fabiocav) [inline src/WebJobs.Script/Description/CSharp/CSharpFunctionSignature.cs]: :+1:

**Comment by davidebbo** [comment]
I'm not familiar with the code object model, but the logic seems sound.

**Comment by fabiocav** [comment]
@davidebbo addressed your feedback and simplified the condition check to decide if we need to restart the host to make it clearer.

**Pull request merged**